### ISSUE TITLE
Cyclic dependency in module 'Darwin': Darwin -> Foundation -> CoreFoundation -> Darwin

### ISSUE THREAD

**Issue by sketchydroide** [google/j2objc#709]
Hi,

I'm having a cyclic dependency on a translated Joda-Convert library, I've already tried to clear and purge all the caches I had I could find about this project, but I still get the same error.

This was after trying to update from j2objc 0.9.8.2.1 to j2objc 1.0.1, I updated and converted all the dependencies, and submodules. 

Also might be important, is that if I try to build on my device, it builds and runs just fine, but when I try to use Analyse instead of Run, or try to Run on the simulator, I get the same compiling issue.

Let me know if you can help me with this, or need any further information.

Thanks,
Andre@laundrapp

**Comment by geoffreywiseman** [comment]
This bit me recently, and I can see other people with the same issue, always from a j2objc dependency:

- https://groups.google.com/forum/#!msg/j2objc-discuss/Q53w8QB5ngU/XlYm-5QPh2AJ
- https://groups.google.com/forum/#!searchin/j2objc-discuss/cyclic$20dependency/j2objc-discuss/jEmhq0ntrvA/mynNh09bW04J

Like many of the other post authors, I had a project building with a j2objc dependency and then, suddenly, out of the blue, this cyclic dependency error appeared.

**Comment by gistya** [comment]
@geoffreywiseman you just saved my life. It was only happening in my unit tests build target. When I made the headers include non-recursive then boom, no more problems. And like you are saying, it just randomly showed up one Monday, like, "How was your weekend?" 👍

**Comment by geoffreywiseman** [comment]
@gistya This cost me a couple hours of fiddling and then reading before I found the fix, so I was hoping it would help someone else. Glad to hear it did. ;)

**Comment by guilhermesgb** [comment]
I'm also getting this kind of issue. The translating steps of j2objc work successfully. But compiling the generated files isn't working as expected.

```
Cyclic dependency in module 'Foundation': Foundation -> CoreFoundation -> Darwin -> Foundation
J2ObjC_common.h
```

![screen shot 2017-03-03 at 15 21 48](https://cloud.githubusercontent.com/assets/465808/23563405/5191ce7c-0025-11e7-901e-c853febd0cf2.png)

But unfortunately, trying out the suggestions found here and elsewhere on the google discussion group didn't work for me.

### Relevant *Build Settings* (all paths below are set to *non-recursive*)
**Framework Search Paths**: `${J2OBJC_HOME}/frameworks`
**Header Search Paths**: `${J2OBJC_HOME}/frameworks/JRE.framework/Headers:${J2OBJC_HOME}/frameworks/Guava.framework/Headers:${J2OBJC_HOME}/frameworks/JSR305.framework/Headers:${J2OBJC_HOME}/frameworks/JUnit.framework/Headers`
**User Header Search Paths**: `${J2OBJC_HOME}/include`

### Relevant *Build Rules*
**Java source files using Script**:
```
"${J2OBJC_HOME}/j2objc" --no-package-directories --swift-friendly -d ${DERIVED_FILES_DIR} -sourcepath "${PROJECT_DIR}/ActiveMobile/SUAL/src:${PROJECT_DIR}/ActiveMobile/SUAL/lib/slf4j-api-1.7.13-sources.jar:${PROJECT_DIR}/ActiveMobile/SUAL/lib/j2objc-annotations-1.3-sources.jar" -classpath "${PROJECT_DIR}/ActiveMobile/SUAL/lib/guava-21.0.jar:${PROJECT_DIR}/ActiveMobile/SUAL/lib/jsr305-1.3.9.jar" -g -Xuse-jdt ${INPUT_FILE_PATH};
```

**Comment by ionut-damian** [comment]
I also have the same problem and the search paths are already set to "non-recursive". Any ideas?

**Comment by tomball** [comment]
This is an Xcode problem, unrelated to j2objc (the only way the j2objc project could "fix" this is to stop generating Objective C, which is the point of the project). [Here's a stackoverflow.com search for "cyclic dependency in module 'Darwin'"](https://stackoverflow.com/search?q=cyclic+dependency+in+module+%27Darwin%27) which found three questions, two with answers.

**Comment by braandl** [comment]
Are there no new ideas to this, or anyone discovered why this appears? So far it seems rather random. I have a project that only compiles when you choose to search headers recursively, then let it fail and compile again without recursive headers :-/

**Comment by tomball** [comment]
Thanks for the solution, @KiwiWilkinson !

**Comment by braandl** [comment]
@KiwiWilkinson This issue still persists. I do only have the includes in the User Header Search Path.

**Comment by KiwiWilkinson** [comment]
@braandl I just fiddled around with the Search Paths until it worked. I'll post my current values here in case they're helpful, but I can't do this until Thursday when I'm back with my development machine.

**Comment by whilethis** [comment]
@KiwiWilkinson @braandl I'm following this because I'm having the same issue. It was working a couple of months ago and I'm guessing the upgrade to Xcode 10.1 broke it. I'm creating static frameworks with my generated code and need a solution ASAP.

**Comment by eliasnemr** [comment]
Hi guys, I've had this cyclic dependency issue for a good month and a half.  I've tried almost everything.   
 I've looked everywhere.. and got 2 replies from Tom Ball.  When it comes to a big project converted from java to Objc using the tool, xcode won't recognize the sub-directories made.  I figured out the best fix is to create your own script using bash.  I did some research and figured out how to recursively find all java files and apply the j2objc script to them and it will output all of your Objc files in the same directory layout. (Keeping a nice layout for you instead of the messy-everything-in-one folder like j2objc script rule on Xcode works).

My build.sh file is as follows, feel free to copy it-

> #!/bin/bash
> #Clean out Objective C folder
> rm -rf /ObjC_Output
> 
> #Move into Java Folder
> cd Java
> 
> #Run j2objc and output everything into ObjC
> j2objc -d ../ObjC_Output $(find . -name "*.java")
> 
> #Jump out of Java folder
> cd ..
> 
> #Delete previous org
> rm -rf /Users/me/Projects/project-name/mobile-app-front-end/app/ios/app/app/org
> 
> #cd to ObjC_Output
> 
> #Copy new ObjC Project files over to the XCode project..
> cp -R ./ObjC_Output/* /Users/me/Projects/project-name/mobile-app-front-end/app/ios/app/app/org

I automatically copy all my files back to my Xcode project.. feel free to do it anyway you want.  Also don't forget to add j2objc to your path.  and run your script as such ./builder.sh in terminal..

Also for Xcode to recognize your objC project folder you must add the top directory of your Xcode project and make it recursive.  

In this case, you don't need to include the j2objc script in Xcode.. Just make sure you include all the default Linker paths, Library search paths and User Header search path as usual (shown in the j2objc documentation).

I've had to pull my hair out for this solution.. So feel free to ask for help if needed. 

Elias

**Comment by KiwiWilkinson** [comment]
Hi Elias,

Your comment about subdirectories is interesting - I have my java in 
subdirectories but didn't have to do this. Admittedly it's 3 directories 
not loads more, but the principle is the same. I don't think I had to 
list the subdirectories to get it to pick up all the files, but I'm not 
absolutely sure as I haven't got XCode with me right now.

Kiwi

**Comment by mungler** [comment]
This is biting me in the ass with a static library project recently... but I found something interesting. Basically, I have a skeleton xcode project for the library (with various non-converted bits of code) into which I merge the converted Java source using a ruby script based on the xcodeproj gem. 

If I set the deployment target of the library to iOS 13.2 (i.e. latest available) everything compiles fine. However, if I try to target an earlier iOS version (our app supports iOS 11 and later, so the library also needs to), I get the cyclic dependency thing.

I'll dig some more and try comparing a working project folder (library targets iOS 13.2) with a non-working one (library targets earlier deployment target) and see if I can figure out what the difference is.
